k8s-scheduler/src/main/java/com/vmware/dcm/ScopedModel.java

```diff
@@ -0,0 +1,129 @@
+    //       divide limitTune to (three) categories according to affinity
+    //       actually tune limitTune
+
+    // TODO: add testcase
+    private double sortWeightCpu = 0.8;
+    private double sortWeightMemory = 0.2;
+
+    // TODO: capture successful schedules history
+
+
+    ScopedModel(final DSLContext conn, final Model model) {
+        this.conn = conn;
+        this.model = model;
+    }
+
+    private Collection<Condition> getWhereClause(final Set<String> nodeSet) {
+        final Collection<Condition> conditions = new ArrayList<>();
+        conditions.add(Tables.SPARE_CAPACITY_PER_NODE.NAME.in(nodeSet));
+
+        return conditions;
+    }
+
+    private SortField<?> getSorting() {
+        return field(sortWeightCpu + " * cpu_remaining + "
+                + sortWeightMemory + " * memory_remaining").desc();
+    }
+
+    private int getLimit(final int cnt) {
+        return (int) Math.ceil(limitTune * cnt);
+    }
+
+    private Set<String> getMatchedNodes () {
+        final Result<?> podNodeSelectorMatches = conn.selectFrom(Tables.POD_NODE_SELECTOR_MATCHES).fetch();
+
+        return podNodeSelectorMatches.intoSet(Tables.POD_NODE_SELECTOR_MATCHES.NODE_NAME);
+    }
+
+    private Set<String> getSpareNodes() {
+        final int podsCnt = conn.fetchCount(selectFrom(Tables.PODS_TO_ASSIGN));
+
+        Result<?> spareNodes = conn.selectFrom(Tables.SPARE_CAPACITY_PER_NODE)
+                .orderBy(getSorting())
+                .limit(getLimit(podsCnt))
+                .fetch();
+
+        return spareNodes.intoSet(Tables.SPARE_CAPACITY_PER_NODE.NAME);
+    }
+
+    public Set<String> getScopedNodes() {
+        Set<String> union = new HashSet<>();
+        Stream.of(
+                getMatchedNodes(),
+                getSpareNodes()
+        ).forEach(union::addAll);
+
+        return union;
+    }
+
+    public Function<Table<?>, Result<? extends Record>> scope() {
+        return (table) -> {
+            if (table.getName().equalsIgnoreCase("spare_capacity_per_node")) {
+
+                // TODO: Used only for log info. Remove if it is expensive to compute.
+                final int sizeUnfiltered = conn.selectFrom(table).fetch().size();
+
+                final Result<?> scopedFetcher = conn.selectFrom(table)
+                        .where(getWhereClause(getScopedNodes()))
+                        .fetch();
+
+                final int sizeFiltered = scopedFetcher.size();
+                LOG.info("solver input size without scope: {}\n" +
+                        "solver input size with scope: {}\n" +
+                        "scope fraction: {}",
+                        sizeUnfiltered, sizeFiltered, (double) sizeFiltered / sizeUnfiltered);
+
+                return scopedFetcher;
+            }
+
+            else {
+                return conn.fetch(table);
+            }
+        };
+    }
+
+    public void setSortWeights(final double weightCpu, final double weightMemory) {
```

Comment:
Do not expose `setSortWeights()`. For now, make the members `private static final`, initialize them the way you currently do (where the field is declared), and rename to "{CPU, MEM}_WEIGHT_DEFAULT`.